### DISCUSSION TITLE
最初の一歩の時間

### DISCUSSION THREAD

**Discussion by Naoki-Hiraoka** [YutaKojio/hrpsys-base#54] [Q&A] [answered]
質問というよりメモです。間違えていたら指摘してください。

footstep_nodes_listの最初と最後の要素は、step_heightが0となり、足を上げません。
https://github.com/YutaKojio/hrpsys-base/blob/a1e7119160a42bdc26c8b44c082981220c5b8abc/rtc/AutoBalancer/GaitGenerator.cpp#L574-L580

そのため、`goVelocity`をすると、最初の一歩ぶんの時間足を上げずに重心だけ動かす処理が行われるので、`goVelocity`を司令してから実際に足を踏み出すまでには一歩分の待ち時間があります。emergency_step時には、すぐに足を踏み出す必要があるので、`emergency_step_time`で0,1,2歩目の時間を変更できるようにしてあり、`emergency_step_time`の0番目の要素を小さく設定しています。

https://github.com/YutaKojio/hrpsys-base/blob/a1e7119160a42bdc26c8b44c082981220c5b8abc/rtc/AutoBalancer/GaitGenerator.cpp#L2132-L2137

https://github.com/YutaKojio/rtmros_tutorials/blob/ba5f3ec4cc8e5d8f4067a93ecf9ba29f3ca02c8c/hrpsys_ros_bridge_tutorials/src/hrpsys_ros_bridge_tutorials/urata_hrpsys_config.py#L359

ただし小さくしすぎると、スライドにあるように、重心を支持脚に十分乗せきれない状態で踏み出すので横に踏み出しがちになります。

**Comment by YutaKojio** [comment, marked as answer]
だいたいあってます．
僕の説明が悪かったですが，最初は一歩時間分両足の真ん中に重心を収束させて，そのあと一歩時間×default_step_time_afterかけて支持足に重心を動かします．（ここでいう一歩時間はどちらもemergency_step_time[0]です．）

step_time=10, default_double_support_ratio=0.4のときは以下のような感じです．
![zmp](https://user-images.githubusercontent.com/11713954/156759070-1f97faee-ef0d-4176-a00b-4c75a54666af.png)
